### PR TITLE
[#2456] Include an optional ID for a tenant's trust anchor

### PR DESCRIPTION
**Pull request by kaniyan**
Extend Device Registry Management API specification to include an optional identifier for a tenant's trust anchor.

### CODE REVIEW

**Comment by sophokles73** [comment]
@kaniyan FMPOV we should also have a first implementation in this PR.

**Comment by kaniyan** [comment]
> @kaniyan FMPOV we should also have a first implementation in this PR.

sure. My intention for creating a PR only with the spec is to ensure that we all have the same understanding before working on any reference implementation. The implementation will follow as consecutive commits.

**Comment by kaniyan** [comment]
> @kaniyan FMPOV we should also have a first implementation in this PR.

@sophokles73 I have now added implementation and tests for the above.

File: tests/src/test/java/org/eclipse/hono/tests/registry/TenantManagementIT.java [hunk 1]
```diff
@@ -158,6 +164,46 @@ public void testAddTenantSucceedsForEmptyBody(final VertxTestContext context) {
...
+
+        getHelper().registry.addTenant(
+                tenantId,
+                tenant,
+                "application/json",
+                HttpURLConnection.HTTP_CREATED)
+                .compose(ok -> getHelper().registry.getTenant(tenantId))
+                .onComplete(context.succeeding(httpResponse -> {
+                    context.verify(() -> {
+                        assertEquals(2,
```
Comment (sophokles73) [inline tests/src/test/java/org/eclipse/hono/tests/registry/TenantManagementIT.java:193]: can we please use *assertThat* instead?

File: tests/src/test/java/org/eclipse/hono/tests/registry/TenantManagementIT.java [hunk 2]
```diff
@@ -158,6 +164,46 @@ public void testAddTenantSucceedsForEmptyBody(final VertxTestContext context) {
...
+                                httpResponse.bodyAsJsonObject().getJsonArray(TenantConstants.FIELD_PAYLOAD_TRUSTED_CA)
+                                        .size());
+                        assertEquals("test-ca", httpResponse.bodyAsJsonObject()
+                                .getJsonArray(RegistryManagementConstants.FIELD_PAYLOAD_TRUSTED_CA).getJsonObject(0)
+                                .getString(RegistryManagementConstants.FIELD_ID));
+                        assertNotNull(httpResponse.bodyAsJsonObject()
```
Comment (sophokles73) [inline tests/src/test/java/org/eclipse/hono/tests/registry/TenantManagementIT.java:199]: IMHO we should use a local variable to store the value of `httpResponse.bodyAsJsonObject()` ...

File: services/device-registry-base/src/main/java/org/eclipse/hono/service/management/tenant/Tenant.java [hunk 3]
```diff
@@ -397,4 +400,55 @@ public boolean hasTrustedCertificateAuthoritySubjectDN(final X500Principal subje
...
+    /**
+     * Asserts that the trust anchor IDs are unique for this tenant and also assigns a unique ID
+     * to each trust anchor for which the ID is not provided.
+     *
+     * @return A reference to this for fluent use.
+     * @throws IllegalArgumentException if the trust anchor IDs are not unique for this tenant.
+     */
+    public final Tenant assertTrustAnchorIdUniquenessAndCreateMissingIds() {
+        assertTrustAnchorIdUniqueness();
+        createMissingTrustAnchorIds();
```
[the remark this answers is not in the document]
Comment (kaniyan) [reply services/device-registry-base/src/main/java/org/eclipse/hono/service/management/tenant/Tenant.java:413]: yes by changing the order we could check the uniqueness of the generated ID too. I saw from a perspective that if there are any duplicates in the ID provided by the user then throwing an error would help the user to take a look and correct it. If there are duplicates in the generated IDs then the problem lies in the server side and he/she cannot do much about it. And I expected the generated IDs to be unique as we use a random UUID for tenants and devices.

File: services/device-registry-base/src/test/java/org/eclipse/hono/service/tenant/AbstractTenantServiceTest.java [hunk 4]
```diff
@@ -408,13 +409,17 @@ default void testGetForCertificateAuthoritySucceeds(final VertxTestContext ctx)
...
         final X500Principal subjectDn = new X500Principal("O=Eclipse, OU=Hono, CN=ca");
 
+        final String trustAnchorId = DeviceRegistryUtils.getUniqueIdentifier();
         final JsonArray expectedCaList = new JsonArray().add(new JsonObject()
+                .put(TenantConstants.FIELD_OBJECT_ID, trustAnchorId)
                 .put(TenantConstants.FIELD_PAYLOAD_SUBJECT_DN, subjectDn.getName(X500Principal.RFC2253))
                 .put(TenantConstants.FIELD_PAYLOAD_PUBLIC_KEY, "NOTAPUBLICKEY".getBytes(StandardCharsets.UTF_8))
                 .put(TenantConstants.FIELD_AUTO_PROVISIONING_ENABLED, false));
 
+        System.out.println(expectedCaList.encodePrettily());
```
[the remark this answers is not in the document]
Comment (kaniyan) [reply services/device-registry-base/src/test/java/org/eclipse/hono/service/tenant/AbstractTenantServiceTest.java:419]: Good catch. I used it for testing and mistakenly didn't remove it.

**Comment by kaniyan** [comment]
@sophokles73 I have made changes in accordance with the above comments and pushed those changes.

File: services/device-registry-base/src/main/java/org/eclipse/hono/service/management/tenant/Tenant.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sophokles73) [reply services/device-registry-base/src/main/java/org/eclipse/hono/service/management/tenant/Tenant.java:413]: The problem remains that we cannot be sure that generated UUIDs do not match any of the client provided IDs unless we check explicitly ...

File: site/documentation/content/api/management/device-registry-v1.yaml [hunk 5]
```diff
@@ -742,6 +742,14 @@ components:
          type: object
          additionalProperties: false
          properties:
+            "id":
+               type: string
+               required: false
+               description: |
+                  The identifier of the trust anchor. This identifier must be unique
+                  among a tenant's trust anchors. If this property is not provided,
+                  then the device registry assigns a unique identifer.
```
Comment (sophokles73) [inline site/documentation/content/api/management/device-registry-v1.yaml:751]: ... identifier.

File: tests/src/test/java/org/eclipse/hono/tests/registry/TenantManagementIT.java [hunk 6]
```diff
@@ -158,6 +164,44 @@ public void testAddTenantSucceedsForEmptyBody(final VertxTestContext context) {
...
+        getHelper().registry.addTenant(
+                tenantId,
+                tenant,
+                "application/json",
+                HttpURLConnection.HTTP_CREATED)
+                .compose(ok -> getHelper().registry.getTenant(tenantId))
+                .onComplete(context.succeeding(httpResponse -> {
+                    context.verify(() -> {
+                        final JsonObject response = httpResponse.bodyAsJsonObject();
+                        assertThat(response.getJsonArray(TenantConstants.FIELD_PAYLOAD_TRUSTED_CA).size()).isEqualTo(2);
```
Comment (sophokles73) [inline tests/src/test/java/org/eclipse/hono/tests/registry/TenantManagementIT.java:194]: how about `assertThat().hasSize()`?

File: services/device-registry-base/src/main/java/org/eclipse/hono/service/management/tenant/Tenant.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kaniyan) [reply services/device-registry-base/src/main/java/org/eclipse/hono/service/management/tenant/Tenant.java]: @sophokles73 I have made some changes to ensure that the generated UUIDs are unique. This is done by verifying if the generated id already exists, if so then a new id is generated until a unique id is available. Then that generated id is used as a trust anchor's id. Would you mind taking a look again?

**Review by sophokles73 — APPROVED** [review]
LGTM

**Pull request merged**